src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs

```diff
@@ -0,0 +1,320 @@
+        /// an update handle, and can not be used for further updates.
+        /// Throws an InvalidOperationException if Update() was already called.
+        /// </summary>
+        public void Update()
+        {
+            if (hUpdate == IntPtr.Zero)
+            {
+                ThrowExceptionForInvalidUpdate();
+            }
+
+            try
+            {
+                if (!EndUpdateResource(hUpdate, false))
+                {
+                    ThrowExceptionForLastWin32Error();
+                }
+            }
+            finally
+            {
+                hUpdate = IntPtr.Zero;
+            }
+        }
+
+
+        private bool EnumTypesCallback(IntPtr hModule, IntPtr lpType, IntPtr lParam)
+        {
+            var enumNamesCallback = new EnumResNameProc(EnumNamesCallback);
+            if (!EnumResourceNames(hModule, lpType, enumNamesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumNamesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, IntPtr lParam)
+        {
+            var enumLanguagesCallback = new EnumResLangProc(EnumLanguagesCallback);
+            if (!EnumResourceLanguages(hModule, lpType, lpName, enumLanguagesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumLanguagesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, ushort wLang, IntPtr lParam)
+        {
+            IntPtr hResource = FindResourceEx(hModule, lpType, lpName, wLang);
+            if (hResource == IntPtr.Zero)
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            // hResourceLoadedq is just a handle to the resource,
+            // which can be used to get the resource data
+            IntPtr hResourceLoaded = LoadResource(hModule, hResource);
+            if (hResourceLoaded == IntPtr.Zero)
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            // This doesn't actually lock memory. It just retrieves a
+            // pointer to the resource data. The pointer is valid
+            // until the module is unloaded.
+            IntPtr lpResourceData = LockResource(hResourceLoaded);
+            if (lpResourceData == IntPtr.Zero)
+            {
+                throw new Exception("Failed to lock resource.");
+             }
+
+            if (!UpdateResource(hUpdate, lpType, lpName, wLang, lpResourceData, SizeofResource(hModule, hResource)))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private static void ThrowExceptionForInvalidUpdate()
+        {
+            throw new InvalidOperationException("Update handle is invalid. This instance may not be used for further updates.");
```

Comment:
What is our localization story for this code base? Exception messages are normally localized.

Reply:
I wasn't sure if it makes sense to take up NETSDK error codes for these exceptions since they should only fire if someone tries to use ResourceUpdater incorrectly. Some of the other non-localized exceptions are implementation details of the updater that could change if we make it possible to customize a windows apphost from unix. @nguerrera do you have a recommendation?

Reply:
We have a few error messages that are generally not possible without a bug, and we've just gone ahead and given them codes and let them be translated. I think it would be fine to do that here too.

---
In reply to: [219338566](https://github.com/dotnet/sdk/pull/2545#discussion_r219338566) [](ancestors = 219338566)

Reply:
Went ahead and made them into localized strings.